rfd/0061-tsh-aliases.md

```diff
@@ -0,0 +1,166 @@
+```
+
+## Configuration and examples
+
+Users will define aliases in their `tsh` configuration file which is kept
+in `$TELEPORT_HOME/config/config.yaml` using the following syntax:
+
+```yaml
+aliases:
+    "<alias>": "<command>"
+```
+
+The `<alias>` can only be a top-level subcommand. In other words, users can
+define a new `tsh mycommand` alias but cannot define `tsh my command` command.
+
+A few notes regarding the `<command>`:
+
+- Unlike `git` aliases explored above, proposal is to always require the binary
+  name instead of defaulting to `tsh` and requiring special syntax otherwise.
+- The `<command>` will be fed to `exec.Command` for execution. To chain multiple
+  commands, use pipes, etc. users will explicitly use `bash -c`.
+- The `<command>` will require `$@`, `$1`, etc. to reference the arguments
+  provided to the alias command.
+
+### Examples
+
+Starting with a very simple example, suppose you're tired of typing `tsh login`
+every morning and want to alias it to something shorter e.g. `tsh l`:
+
+```yaml
+aliases:
+    "l": "tsh login $@"
+```
+
+As a more practical example, the following alias will make `tsh login` default
+to the specific auth connector and username:
+
+```yaml
+aliases:
+    "login": "tsh login --auth=local --user=alice $@"
+```
+
+Similarly, as discussed in the search-based access requests RFD, some users
+may want `tsh ssh` to default to the mode in which it will auto-request access
+to the node upon encountering an access denied error:
+
+```yaml
+aliases:
+    "ssh": "tsh ssh -P $@"
+```
+
+Note: command arguments will be resolved prior to invoking the alias. So in the
+example above, when a user runs `tsh ssh root@node1`, the alias command executed
+will be `tsh ssh -P root@node1`.
+
+### Environment variables
+
+Each alias invocation will set a few environment variables that can be consumed
+within the alias.
+
+To prevent infinite recursion, `TSH_ALIAS` variable will be set indicating the
+command is invoked as an alias. If it's detected, `tsh` won't attempt to expand
+the same command again. For example, `TSH_ALIAS=login`. This will prevent other
+`login` commands from being expanded but will allow using aliases in other
+aliases (see below for an example).
+
+Each alias will also set `TSH` environment variable to the path of the `tsh`
+binary that invoked the alias and `SHELL` environment variable to the default
+shell of the user invoking the command.
+
+### More examples
+
+An alias can also define a custom subcommand that combines multiple commands.
+The following alias will connect to a node within a specific leaf cluster
+using `tsh connect leaf ubuntu@node-1` command:
+
+```yaml
+aliases:
+    "connect": "bash -c '$TSH login $1 && $TSH ssh $2'"
+```
+
+The following alias will list nodes in all clusters:
+
+```yaml
+aliases:
+    "lsall": "bash -c 'for cluster in $($TSH clusters | tail -3 | head -2 | cut -d \' \' -f1); $TSH ls --cluster=$cluster; done'"
```

Comment:
This will work but be pretty brittle, let's not forget to add `tsh ls -R` to allow global search.